src/intel_pstate.c

```diff
@@ -0,0 +1,148 @@
+/**
+ *  intel_pstate.c - automatically control turbo frequency for MacBook Pro
+ *  Copyright (C) 2021 Gokturk Yuksek <gokturk@gentoo.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <dirent.h>
+#include <errno.h>
+
+#include "intel_pstate.h"
+#include "util.h"
+#include <syslog.h>
+
+static int read_int(FILE *fp, int *val)
+{
+    char buf[4]; /* Maximum we can read in this module is '100\0' */
+    ssize_t ret;
+
+    ret = pread(fileno(fp), buf, sizeof(buf), 0);
+    if (ret >= 0) {
+        buf[ret] = '\0';
+        *val = atoi(buf);
+        return 0;
+    } else {
+        return (int)ret;
+    }
+}
+
+static inline int write_str(FILE *fp, char *str, int len)
+{
+    ssize_t ret;
+    ret = (int)pwrite(fileno(fp), str, len, 0);
+    if (ret == len)
+        return 0;
+    else
+        return -1;
+}
+
+static int write_int(FILE *fp, int val)
+{
+    char buf[4]; /* Maximum we can read in this module is '100\0' */
+    int ret;
+
+    ret = snprintf(buf, sizeof(buf), "%d", val);
+    if (ret < 0)
+        return ret;
+
+    return write_str(fp, buf, ret);
+}
+
+int intel_pstate_init(t_intel_pstate *intel_pstate)
+{
+    const char *path_max_perf_pct = INTEL_PT_PATH "/max_perf_pct";
+    const char *path_min_perf_pct = INTEL_PT_PATH "/min_perf_pct";
+    FILE *fp;
+    int err;
+
+    fp = fopen(path_max_perf_pct, "w+");
+    if (!fp)
+        return -1;
+    /* Save the initial value of max_perf_pct, to restore it on exit */
+    err = read_int(fp, &intel_pstate->preserved_max_perf_pct);
```

Comment:
This is not a performance-critical path -- can we use the existing `read_value` or `fscanf("%d", &intel_pstate->preserved_max_perf_pct)`?  Using the standard library avoids issues with `EINTR` and generally makes the code easily to understand.

Reply:
Not a problem. Will switch to `fscanf`.